crates/burn-train/src/learner/train_val.rs

```diff
@@ -144,14 +144,40 @@ impl<LC: LearnerComponents> Learner<LC> {
             None => 1,
         };
 
-        for epoch in starting_epoch..self.num_epochs + 1 {
-            let epoch_train = TrainEpoch::new(
-                dataloader_train.clone(),
-                epoch,
-                self.num_epochs,
-                self.grad_accumulation,
-            );
+        // `MultiDevicesTrainStep` has one worker per device, so we use a fixed device strategy
+        // for each (worker) data loader. This matches the expected device on the worker, so we
+        // don't have to move the data between devices.
+        let dataloaders_train = if self.devices.len() > 1 {
+            let num_splits = self.devices.len();
+            let num_items = dataloader_train.num_items();
+            let mut dataloaders = Vec::with_capacity(num_splits);
+
+            let mut start = 0;
+            let step = num_items / num_splits;
+            for i in 0..num_splits {
+                let end = if i == (num_splits - 1) {
+                    num_items
+                } else {
+                    start + step
+                };
+                let mut dataloader = dataloader_train.slice(start, end);
+                dataloader.set_device(self.devices[i].clone());
+                dataloaders.push(Arc::from(dataloader));
+                start = end;
+            }
+            dataloaders
+        } else {
+            vec![dataloader_train]
+        };
```

Comment:
I would extract that into a new function, maybe called: `split_dataloaders(dataloader, &devices)`